drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2388 @@
+#include <zephyr/drivers/clock_control/renesas_cpg_mssr.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/cache.h>
+#include <zephyr/drivers/regulator.h>
+
+#include "rcar_mmc_registers.h"
+
+#define PINCTRL_STATE_UHS PINCTRL_STATE_PRIV_START
+
+/**
+ * @note we don't need any locks here, because SDHC subsystem cares about it
+ */
+
+LOG_MODULE_REGISTER(rcar_mmc, CONFIG_LOG_DEFAULT_LEVEL);
+
+#define MMC_POLL_FLAGS_TIMEOUT_US 100000
+#define MMC_POLL_FLAGS_ONE_CYCLE_TIMEOUT_US 1
+#define MMC_BUS_CLOCK_FREQ 800000000
+/*
+ * SD/MMC clock for Gen3/Gen4 R-car boards can't be equal to 208 MHz,
+ * but we can run SDR104 on lower frequencies:
+ *    "SDR104: UHS-I 1.8V signaling, Frequency up to 208 MHz"
+ * so according to SD card standard it is possible to use lower frequencies,
+ * and we need to pass check of frequency in sdmmc in order to use sdr104 mode.
+ * This is the reason why it is needed this correction.
+ */
+#define MMC_MAX_FREQ_CORRECTION 8000000
+
+#ifdef CONFIG_RCAR_MMC_DMA_SUPPORT
+#define ALIGN_BUF_DMA __aligned(CONFIG_SDHC_BUFFER_ALIGNMENT)
+#else
+#define ALIGN_BUF_DMA
+#endif
+
+/**
+ * @brief Renesas MMC host controller driver data
+ *
+ */
+struct mmc_rcar_data {
+	DEVICE_MMIO_RAM; /* Must be first */
+	struct sdhc_io host_io;
+	struct sdhc_host_props props;
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	struct k_sem irq_xref_fin;
+#endif
+
+	uint8_t ver;
+	/* in bytes, possible values are 2, 4 or 8 */
+	uint8_t width_access_sd_buf0;
+	uint8_t ddr_mode;
+	uint8_t dma_support;
+	uint8_t restore_cfg_after_reset;
+	uint8_t is_last_cmd_app_cmd; /* ACMD55 */
+
+#ifdef CONFIG_RCAR_MMC_SCC_SUPPORT
+	uint8_t manual_retuning;
+	uint8_t tuning_buf[128] ALIGN_BUF_DMA;
+#endif /* CONFIG_RCAR_MMC_SCC_SUPPORT */
+	uint8_t can_retune;
+};
+
+/**
+ * @brief Renesas MMC host controller driver configuration
+ */
+struct mmc_rcar_cfg {
+	DEVICE_MMIO_ROM; /* Must be first */
+	struct rcar_cpg_clk cpg_clk;
+	struct rcar_cpg_clk bus_clk;
+	const struct device *cpg_dev;
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *regulator_vqmmc;
+	const struct device *regulator_vmmc;
+
+	uint32_t max_frequency;
+
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	void (*irq_config_func)(const struct device *dev);
+#endif
+
+	uint8_t non_removable;
+	uint8_t uhs_support;
+	uint8_t mmc_hs200_1_8v;
+	uint8_t mmc_hs400_1_8v;
+	uint8_t bus_width;
+	uint8_t mmc_sdr104_support;
+};
+
+#ifdef CONFIG_RCAR_MMC_SCC_SUPPORT
+static int rcar_mmc_execute_tuning(const struct device *dev);
+static int rcar_mmc_retune_if_needed(const struct device *dev, bool request_retune);
+#endif
+static int rcar_mmc_disable_scc(const struct device *dev);
+
+/**
+ * @brief Read 32-bit Renesas MMC register
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device, used for getting reg base address
+ * @param reg register offset relative to the base address
+ *
+ * @retval register value
+ */
```

Comment:
What are the expectations of using Doxygen comments for driver-internal static functions?
Especially when you comment on something quite common and obvious?

Reply:
I aligned functions' descriptions with one another. As for me, well-documented code is welcomed in open communities. Moreover, a lot of drivers have Doxygen-style comments for driver-internal static functions.

Reply:
> I aligned functions' descriptions with one another. As for me, well-documented code is welcomed in open communities.

`well-documented code !=  comment on something quite common and obvious`

>Moreover, a lot of drivers have Doxygen-style comments for driver-internal static functions.

No. And there is no reason to do same nonsense.

NACK.

Reply:
Is there any documented requirement that prohibits adding Doxygen-style comments for driver-internal static functions, or is it only your own preference?

Reply:
No, but there is an undocumented preference not to do stupid things. Your driver is not a public API, so there is absolutely no reason to use Doxygen comments.